File: haskell/p-28.hs
import qualified Data.List as D
import qualified Data.Map as M

lsort :: [[a]] -> [[a]]
lsort = D.sortBy (\x y -> compare (length x) (length y))

lfsort :: [[a]] -> [[a]]
lfsort xs =
    let ls = map length xs
        gs = D.group $ D.sort ls
        lf = M.fromList  (map (\x -> (head x, length x)) gs)
    in
      D.sortBy (\x y -> compare (M.lookup (length x) lf)
                        (M.lookup (length y) lf)) xs

-- *Main> lfsort ["a", "b", "c", "de", "fg", "ijk"]
-- ["ijk","de","fg","a","b","c"]

File: erlang/p_28.erl
-module(p_28).
-export([lsort/1, lfsort/1]).
-import(p_09, [pack/1]).
-import(maps, [get/2, from_list/1]).


lsort(Xs) ->
    lists:sort(fun(X, Y) ->
		       length(X) < length(Y)
	       end, Xs).

lfsort(Xs) ->
    Ls = lists:map(fun (X) -> length(X) end, Xs),
    Gs = p_09:pack(lists:sort(Ls)),
    Lf = from_list(lists:map(fun([Y|Ys]) ->
					  {Y, length([Y|Ys])}
				  end, Gs)),
    lists:sort(fun (X, Y) ->
		       get(length(X), Lf) < get(length(Y), Lf)
	       end, Xs).


%% 133> p_28:lfsort(["a", "b", "c", "de", "fg", "ijk"]).
%% ["ijk","fg","de","c","b","a"]

